File: serving/frontend/src/api/conversation.js
import { request } from './index.js'

/**
 * Load conversation history for a project.
 *
 * @param {string} projectId
 * @param {{ limit?: number, before?: string }} options
 */
export async function getConversation(projectId, { limit = 50, before = null } = {}) {
  const params = new URLSearchParams({ limit: String(limit) })
  if (before) params.set('before', before)
  return request(`/projects/${encodeURIComponent(projectId)}/conversation?${params}`)
}

/**
 * Persist a message to the server-side conversation log.
 *
 * @param {string} projectId
 * @param {{ type?: string, content: string, metadata?: object }} message
 */
export async function sendMessage(projectId, { type = 'user', content, metadata = {} }) {
  return request(`/projects/${encodeURIComponent(projectId)}/conversation`, {
    method: 'POST',
    body: JSON.stringify({ type, content, metadata }),
  })
}

/**
 * Clear the server-side conversation history for a project.
 *
 * @param {string} projectId
 */
export async function clearConversation(projectId) {
  return request(`/projects/${encodeURIComponent(projectId)}/conversation`, {
    method: 'DELETE',
  })
}

File: serving/api/conversation.py
"""Conversation API endpoints."""

from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query

from api.users import get_current_user_id, get_optional_user_id
from models.conversation import ConversationResponse, SendMessageRequest
from services.conversation_service import get_conversation_service

router = APIRouter(prefix="/projects/{project_id}/conversation", tags=["conversation"])


@router.get("", response_model=ConversationResponse)
async def get_conversation(
    project_id: str,
    limit: int = Query(default=500, ge=1, le=500),
    before: Optional[str] = Query(default=None),
    user_id: Optional[str] = Depends(get_optional_user_id),
):
    """Get conversation messages for a project."""
    service = get_conversation_service()
    if not service:
        raise HTTPException(status_code=503, detail="Conversation service not available")

    messages, total, has_more = await service.get_messages(
        project_id=project_id,
        limit=limit,
        before=before,
    )

    return ConversationResponse(messages=messages, total=total, has_more=has_more)


@router.post("", status_code=201)
async def send_message(
    project_id: str,
    body: SendMessageRequest,
    user_id: str = Depends(get_current_user_id),
):
    """Send a message to a project conversation."""
    service = get_conversation_service()
    if not service:
        raise HTTPException(status_code=503, detail="Conversation service not available")

    # Get user display name
    from services.user_service import get_user_service
    user_service = get_user_service()
    display_name = "Unknown"
    if user_service:
        user = await user_service.get_user(user_id)
        if user:
            display_name = user.username

    msg = await service.add_message(
        project_id=project_id,
        user_id=user_id,
        display_name=display_name,
        type=body.type,
        content=body.content,
        metadata=body.metadata,
    )

    return msg.model_dump()


@router.delete("", status_code=204)
async def clear_conversation(
    project_id: str,
    user_id: str = Depends(get_current_user_id),
):
    """Clear conversation for a project."""
    service = get_conversation_service()
    if not service:
        raise HTTPException(status_code=503, detail="Conversation service not available")

    await service.clear(project_id)
